contracts/VirtualSwap/Bridge.sol

```diff
@@ -0,0 +1,619 @@
+            tokenFromIndex,
+            v.synthIndex,
+            tokenInAmount,
+            0,
+            block.timestamp
+        );
+
+        IERC20 mediumSynth = IERC20(_getSynthAddress(swap));
+        v.synthSwapper = new SynthSwapper();
+        mediumSynth.transfer(address(v.synthSwapper), v.mediumSynthAmount);
+
+        // Swap synths via Synthetix network
+        v.destSynthAmount = v.synthSwapper.swapSynth(
+            SYNTHETIX,
+            _getSynthKey(swap),
+            v.mediumSynthAmount,
+            synthOutKey
+        );
+        require(v.destSynthAmount >= minAmount, "Insufficient output");
+
+        // Add the synthswapper to the pending settlement list
+        v.itemId = _addToPendingSynthSettlementList(
+            PendingSynthSettlement(
+                v.synthSwapper,
+                synthOutKey,
+                recipient,
+                false
+            )
+        );
+
+        // Emit TokenToSynth event with relevant data
+        emit TokenToSynth(
+            msg.sender,
+            swap,
+            v.tokenFrom,
+            tokenInAmount,
+            v.destSynthAmount,
+            v.itemId
+        );
+
+        return (v.itemId);
+    }
+
+    function calcSynthToToken(
+        ISwap swap,
+        bytes32 synthInKey,
+        uint8 tokenToIndex,
+        uint256 synthInAmount
+    ) external view returns (uint256) {
+        IExchangeRates exchangeRates =
+            IExchangeRates(SYNTHETIX_RESOLVER.getAddress(EXCHANGE_RATES_NAME));
+
+        uint8 mediumSynthIndex = _getSynthIndex(swap);
+        bytes32 mediumSynthKey = _getSynthKey(swap);
+        require(synthInKey != mediumSynthKey, "use normal swap");
+
+        uint256 expectedMediumSynthAmount =
+            exchangeRates.effectiveValue(
+                synthInKey,
+                synthInAmount,
+                mediumSynthKey
+            );
+
+        return
+            swap.calculateSwap(
+                mediumSynthIndex,
+                tokenToIndex,
+                expectedMediumSynthAmount
+            );
+    }
+
+    // Swaps any synth to a token that Saddle's pools support
+    function synthToToken(
+        ISwap swap,
+        bytes32 synthInKey,
+        uint8 tokenToIndex,
+        uint256 synthInAmount,
+        uint256 minAmount,
+        address recipient
+    ) external returns (uint256) {
+        SynthToTokenInfo memory v = SynthToTokenInfo(0, SynthSwapper(0), 0);
+
+        if (recipient == address(0)) {
+            recipient = msg.sender;
+        }
+
+        // Recieve synth from the user
+        IERC20 synthFrom =
+            IERC20(Target(SYNTHETIX_RESOLVER.getSynth(synthInKey)).proxy());
+        synthFrom.safeTransferFrom(msg.sender, address(this), synthInAmount);
+
+        v.mediumSynthKey = _getSynthKey(swap);
+        v.mediumSynthIndex = _getSynthIndex(swap);
+        require(
+            synthInKey != v.mediumSynthKey,
+            "synth is supported via normal swap"
+        );
```

Comment:
Should this check happen before the `safeTransferFrom` per checks-effects-interactions?